include: "rules/common.smk"
include: "rules/convert.smk"
include: "rules/derep.smk"
include: "rules/swarm.smk"


(SAMPLES,) = glob_wildcards("reads/final/{sample}.fastq")


rule all:
    input:
        "data/swarm_stats_without_singletons.tsv",
        "data/swarm_cluster_representatives_without_singletons.fna",


rule stats:
    input:
        expand("reads/final/{sample}.swarm_stats_pretty.tsv", sample=SAMPLES),
    output:
        "data/swarm_stats.tsv",
    shell:
        """
        qsv cat rowskey --group fstem --group-name sample --output {output} {input}
        sed -i "s/\\.swarm_stats_pretty//" {output}
        """


rule swarm_stats_without_singletons:
    input:
        "data/swarm_stats.tsv",
    output:
        "data/swarm_stats_without_singletons.tsv",
    shell:
        """
        qsv search --invert-match --select total_abundance "^1$" --output {output} {input}
        """


rule swarm_stats_without_singletons_ids:
    input:
        "data/swarm_stats_without_singletons.tsv",
    output:
        temp("data/swarm_non_singletons.tsv"),
    shell:
        """
        qsv select seed {input} | qsv sort --unique | qsv behead --output {output}
        """


rule representatives:
    input:
        expand("reads/final/{sample}.cluster_representatives.fna", sample=SAMPLES),
    output:
        "data/swarm_cluster_representatives.fna",
    shell:
        "seqkit rmdup --quiet --by-seq {input}"
        " | seqkit replace --pattern ';.+' --replacement ''"
        " | seqkit sort --quiet --line-width 0"
        " > {output}"


rule swarm_cluster_representatives_without_singletons:
    input:
        fasta="data/swarm_cluster_representatives.fna",
        ids="data/swarm_non_singletons.tsv",
    output:
        fasta="data/swarm_cluster_representatives_without_singletons.fna",
        index=temp("data/swarm_cluster_representatives_without_singletons.fna.fai"),
    shell:
        "seqkit faidx --quiet --line-width 0 --infile-list {input.ids} {input.fasta} > {output}"
